File: content/snippets/scala3/language_basics/trees_and_graphs.scala
import scala.collection.mutable

object TreesAndGraphs:
  case class TreeNode(value: Int, var left: Option[TreeNode] = None, var right: Option[TreeNode] = None)

  def insert(root: Option[TreeNode], value: Int): TreeNode = root match {
    case None => TreeNode(value)
    case Some(node) =>
      if (value < node.value) node.left  = Some(insert(node.left,  value))
      else                    node.right = Some(insert(node.right, value))
      node
  }

  def inOrder(root: Option[TreeNode]): List[Int] = root match {
    case None       => Nil
    case Some(node) => inOrder(node.left) ::: node.value :: inOrder(node.right)
  }

  def bfs(root: Option[TreeNode]): List[Int] =
    val queue  = mutable.Queue[TreeNode]()
    val result = mutable.ListBuffer[Int]()
    root.foreach(queue.enqueue)
    while (queue.nonEmpty) {
      val node = queue.dequeue()
      result += node.value
      node.left.foreach(queue.enqueue)
      node.right.foreach(queue.enqueue)
    }
    result.toList

  @main def main(): Unit =
    var root: Option[TreeNode] = None
    for (v <- List(5, 3, 7, 1, 4)) root = Some(insert(root, v))
    println(s"in-order: ${inOrder(root)}")  // List(1, 3, 4, 5, 7)
    println(s"bfs:      ${bfs(root)}")      // List(5, 3, 7, 1, 4)

File: content/snippets/go/language_basics/trees_and_graphs.go
package main

import "fmt"

type TreeNode struct {
	value       int
	left, right *TreeNode
}

func insert(root *TreeNode, value int) *TreeNode {
	if root == nil {
		return &TreeNode{value: value}
	}
	if value < root.value {
		root.left = insert(root.left, value)
	} else {
		root.right = insert(root.right, value)
	}
	return root
}

func inOrder(root *TreeNode) {
	if root == nil {
		return
	}
	inOrder(root.left)
	fmt.Print(root.value, " ")
	inOrder(root.right)
}

func bfs(root *TreeNode) {
	if root == nil {
		return
	}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		fmt.Print(node.value, " ")
		if node.left != nil {
			queue = append(queue, node.left)
		}
		if node.right != nil {
			queue = append(queue, node.right)
		}
	}
}

func main() {
	var root *TreeNode
	for _, v := range []int{5, 3, 7, 1, 4} {
		root = insert(root, v)
	}
	fmt.Print("in-order: ")
	inOrder(root)
	fmt.Println() // 1 3 4 5 7
	fmt.Print("bfs:      ")
	bfs(root)
	fmt.Println() // 5 3 7 1 4
}
